### DISCUSSION TITLE
GoB v3_5_91

### DISCUSSION THREAD

**Discussion by kromar** [JoseConseco/GoB#356] [General]
**Fixed** 
- Manual import is now working every time the button is pressed if GoZ models exist on drive
- Enabling the import button would trigger a import instead of waiting for the next input from Zbursh. This has been fixed and the first import will happen once something is sent out from Zbrush.
- When using the flip axis options the object would be flipped, this is no longer the case and a issue with flipping normals has been fixed as well. 
<img src="https://user-images.githubusercontent.com/1472884/181602825-1c76b6a3-b4e9-4b66-ab5b-ce311cdf633f.png" width="200" />
- polygroups and vertex groups did no longer import into blender

**Improved**
- When importing materials from textures it could happen that nodes were missing or not created. This has been fixed and the whole node tree will be created. Further improvements on this will follow.
- reduced import time of polygroups

<hr /><em>This discussion was created from the release <a href='https://github.com/JoseConseco/GoB/releases/tag/v3_5_91'>GoB v3_5_91</a>.</em>

**Comment by sidney-eliot** [comment]
Found a uncatched error. Happens when exporting object from blender in edit mode

```
location: <unknown location>:-1
Error: Python: Traceback (most recent call last):
  File "C:\Users\Epicrex\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\GoB-3_5_91\GoB.py", line 1394, in invoke
    bpy.ops.scene.gob_export(as_tool=as_tool)
  File "E:\ProgramFiles\Steam\steamapps\common\Blender\3.0\scripts\modules\bpy\ops.py", line 132, in __call__
    ret = _op_call(self.idname_py(), None, kw)
RuntimeError: Error: Python: Traceback (most recent call last):
  File "C:\Users\Epicrex\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\GoB-3_5_91\GoB.py", line 1309, in execute
    self.exportGoZ(PATH_GOZ, context.scene, obj, f'{PATH_PROJECT}')
  File "C:\Users\Epicrex\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\GoB-3_5_91\GoB.py", line 762, in exportGoZ
    me = apply_modifiers(obj)
  File "C:\Users\Epicrex\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\GoB-3_5_91\GoB.py", line 1870, in apply_modifiers
    result = bmesh.ops.triangulate(bm, faces=tf)
TypeError: faces: 0 BMFace has been removed

location: E:\ProgramFiles\Steam\steamapps\common\Blender\3.0\scripts\modules\bpy\ops.py:132



location: <unknown location>:-1
```


Added to that exporting from edit mode should be allowed

**Reply by kromar** [reply]
here is a fix for it, i have not tested it to much so let me know if you encounter any further issues. https://github.com/kromar/blender_GoB/tree/fix-triangulation-error

**Comment by sidney-eliot** [comment]
I am also noticing that when I click GOZ in ZBrush and then activate import in Blender, the object isn't sent to Blender. I have to click GOZ again for it to be sent. Not sure if this intended

**Reply by kromar** [reply]
this is a change i made, the reason is that if there are some exported objects from a previous session it would import those which is not what i would expect. There is always the manual import button in such a case which you encounter.

**Reply by sidney-eliot** [reply]
Couldn't you fix that previous session issue by emptying the link folder after x criteria is met

**Reply by kromar** [reply]
Im not sure, maybe something like that would be possible if there is some timer that tells me how long a blender session has been running... i will check if i can come up with something for that.

**Reply by sidney-eliot** [reply]
Do be honest I have never tried the manual import button, don't know what is does

**Reply by kromar** [reply]
whenever you export something from zbrush with GOZ the meshes are stored in a folder, usually the auto import will check the age of a model there and imports if it finds a new one. The manual button will just import whatever is in that folder when you press it.

**Reply by sidney-eliot** [reply]
I was more thinking of application close, but I don't know if you could detect from a blender addon, if both ZBrush and blender are closed. But I think having that import /GoZ button more ergonomic with the drawback of having the possibility of the link folder emptied prematurely is possibly worth it

**Reply by sidney-eliot** [reply]
So the manual button will basically most of the times import way to many objects?

**Reply by kromar** [reply]
No, just what has been last exported from zbrush.

**Comment by kromar** [comment]
Ah yes, someone else was reporting this but i did not know how to reproduce it. Thanks for reporting it i will see if i can fix the error.

**Reply by sidney-eliot** [reply]
I think that error description of edit mode was false, I was just coincidentally in edit mode, object mode also doesn't work at the moment and gives the same error at export. I think it has something to do with the path. Sending from ZBrush to Blender works but not the other way around

**Reply by sidney-eliot** [reply]
I tried the ..._92 on the triangulation fix branch and it works there again

**Comment by CountQuadula** [comment]
When installing Gob
Is the ZBrush Path supposed to be manually selected?  on install, It was blank. ~ thanks
![zb2022_GoBPath](https://user-images.githubusercontent.com/42452955/181795613-3fb6e013-f106-41cb-a81c-007f665b41ec.jpg)

**Reply by sidney-eliot** [reply]
Yes I noticed this as well, I think in earlier versions it set the path to the default ZBrush install path. Little tip on getting a exe path as fast as possible: Win button, search ZBrush, right click > open file location, right click on ZBrush > properties and there's the path (or use a file search tool like listary)

**Reply by kromar** [reply]
if the path is empty the addon will search for the most recent zbrush installation and will use and add it as the default path for future exports, so there is no need to specify that path unless you have updated your zbrush version which is not recognized at the moment.

**Reply by sidney-eliot** [reply]
Ok but this has to be clear to the user, maybe the field says default path and when one click into it it empties so one can enter a custom path. Or maybe you put something in the title for that field in this case "ZBrush Path"

**Reply by kromar** [reply]
Im not sure i fully understand your proposal to make this more clear to the user. Would renaming the field and having a more detailed tooltip enough or do you propose to change some logic for this input?
![image](https://user-images.githubusercontent.com/1472884/182907162-928c80b4-32cf-4647-b29a-ebdafa72c232.png)

**Comment by sidney-eliot** [comment]
I'm working on a character project right now and my tool has 90 SubTools all of which are 22mil polys in total, exporting one object from blender to ZBrush takes about 11 sec. There are allot of SubTools that will however still have to done and the poly count will also increase allot, so this export time will get more and more over time. Is this in line with what you are getting time wise? I think it might also be so slow because I'm using Blender and ZBrush on my HDD however the project path (where the temp files are stored) is on my SSD.

**Reply by kromar** [reply]
That sounds about the current performance, there is a option in the debug section to show you how long certain functions take which might allow you t disable certain things you dont need to transfer. Im aware that this is only a workaround to gain a bit of speedup until i find some more time to improve the import/export speed further. There is certainly a big potential to gain some speedups but its not a trivial task and will take some time to implement.
![image](https://user-images.githubusercontent.com/1472884/182904571-4ec549ed-3eaa-40c6-85f8-461b74dd790d.png)

**Reply by sidney-eliot** [reply]
I don't understand the time value, I did some optimizing and I'm now at 142.41... ms (for the total time) which is a fraction of a second and it definitely takes allot longer than that. And for importing I'm also getting a super small value: 180 ms
